mmdet/models/dense_heads/yolo_head.py

```diff
@@ -215,200 +215,68 @@ def get_bboxes(self,
-                    cls_pred = cls_pred[batch_inds, topk_inds, :]
-                    conf_pred = conf_pred[batch_inds, topk_inds]
-            # Save the result of current scale
-            multi_lvl_bboxes.append(bbox_pred)
-            multi_lvl_cls_scores.append(cls_pred)
-            multi_lvl_conf_scores.append(conf_pred)
-
-        # Merge the results of different scales together
-        batch_mlvl_bboxes = torch.cat(multi_lvl_bboxes, dim=1)
-        batch_mlvl_scores = torch.cat(multi_lvl_cls_scores, dim=1)
-        batch_mlvl_conf_scores = torch.cat(multi_lvl_conf_scores, dim=1)
-
-        # Replace multiclass_nms with ONNX::NonMaxSuppression in deployment
-        if torch.onnx.is_in_onnx_export() and with_nms:
-            from mmdet.core.export import add_dummy_nms_for_onnx
-            conf_thr = cfg.get('conf_thr', -1)
-            score_thr = cfg.get('score_thr', -1)
-            # follow original pipeline of YOLOv3
-            if conf_thr > 0:
-                mask = (batch_mlvl_conf_scores >= conf_thr).float()
-                batch_mlvl_conf_scores *= mask
-            if score_thr > 0:
-                mask = (batch_mlvl_scores > score_thr).float()
-                batch_mlvl_scores *= mask
-            batch_mlvl_conf_scores = batch_mlvl_conf_scores.unsqueeze(
-                2).expand_as(batch_mlvl_scores)
-            batch_mlvl_scores = batch_mlvl_scores * batch_mlvl_conf_scores
-            max_output_boxes_per_class = cfg.nms.get(
-                'max_output_boxes_per_class', 200)
-            iou_threshold = cfg.nms.get('iou_threshold', 0.5)
-            # keep aligned with original pipeline, improve
-            # mAP by 1% for YOLOv3 in ONNX
-            score_threshold = 0
-            nms_pre = cfg.get('deploy_nms_pre', -1)
-            return add_dummy_nms_for_onnx(
-                batch_mlvl_bboxes,
-                batch_mlvl_scores,
-                max_output_boxes_per_class,
-                iou_threshold,
-                score_threshold,
-                nms_pre,
-                cfg.max_per_img,
-            )
-
-        if with_nms and (batch_mlvl_conf_scores.size(0) == 0):
+        num_imgs = len(img_metas)
+        featmap_sizes = [pred_map.shape[-2:] for pred_map in pred_maps]
+
+        mlvl_anchors = self.anchor_generator.grid_anchors(
+            featmap_sizes, pred_maps[0].device)
+        flatten_preds = []
+        flatten_strides = []
+        for pred, stride in zip(pred_maps, self.featmap_strides):
+            pred = pred.permute(0, 2, 3, 1).reshape(num_imgs, -1,
+                                                    self.num_attrib)
+            pred[..., :2].sigmoid_()
+            flatten_preds.append(pred)
+            flatten_strides.append(
+                pred.new_tensor(stride).expand(pred.size(1)))
+
+        flatten_preds = torch.cat(flatten_preds, dim=1)
+        flatten_bbox_preds = flatten_preds[..., :4]
+        flatten_objectness = flatten_preds[..., 4].sigmoid()
```

Comment:
Why can't the sigmoid be unified in the front but calculated separately?

Reply:
It seems that using inplace sigmoid is a little bit slower than create a new tensor. Maybe because of the memory continuity.